design-proposals/e2e-waitfor-tooling.md

```diff
@@ -0,0 +1,154 @@
+Event driven E2E Testing
+=
+
+**Authors**: [Edward Haas](https://github.com/eddev)
+
+# Summary
+
+This proposal is suggesting the usage of events over polling the cluster
+resources & objects in e2e tests.
+
+# Motivation
+
+Kubevirt has an impressive amount of e2e test, covering a large amount
+scenarios and features.
+
+The scenarios that the tests are exercising, frequently use asynchronous
+API operations against the Kubevirt cluster. In most cases, there is a
+need to *wait* for the (overall) operation to complete (be examining
+the current state) and only then proceed with other steps.
+
+Waiting for a state of an object (e.g. VMI) is currently implemented
+in the e2e tests through polling repeatedly its state.
+This polling strategy has one clear pro, it is simple.
+On the other hand, there are several cons:
+- Loads the API server (depending on the number of retries, frequency of
+  retries, concurrent running tests,  mount of such "waits").
+  This load is also dependent on the resources the API Server has in the
+  test cluster.
+- Intermediate states may be missed. The sampling of the state may just
+  miss some conditions, just because it happened very fast.
+  Mitigating this usually involves the need to increase the frequency of
+  such sampling, which again raises the load issue.
+
+Therefore, one can claim that using polling to wait for a condition to
+be met is by definition flaky and inaccurate.
+
+Fortunately, Kubernetes core strategy is to use events to distribute
+facts about the objects/resources state.
+Production application that integrate with Kubernetes are expected to
+use events (e.g. watchers, informers) and not polling.
+
+The e2e tests can be classified as applications that integrate with the
+Kubernetes cluster, and therefore, could (and should) use the same
+patterns expected from production applications.
+
+## Goals
+
+- Use event driven approach to wait for a condition. Encourage it over
+  the polling approach.
+- Increase e2e test stabilization.
```

Comment:
to be clear, this question is not an argument against using an event watching pattern

I understand that watching is more efficient that pooling, but for our test suite, do we have evidence that our existing poll logic is causing instability?

Reply:
@dankenigsberg I am not sure the example is relevant for this discussion. (Side note: The test could use eventually as we expect that eventually, we will fail to create the VM after deleting the CDI)

Reply:
yeah, in this specific case I don't think we have any objects to watch in order to ensure the kubevirt components have transitioned to not using CDI anymore. Reporting that on an object (like the kv object as a condition) might be a good improvement in the future. 

That isn't exactly supporting evidence for instability related to polling vs event watching within the context of using informers vs GET requests though.

Reply:
Yeah, my URL is relevant only in a very convoluted way (a feature should be written in such a way that it can be tested with a watch). A better example (still theoretical, no proof for the assertion in the doc) is every time we use `Consistently` in a test. Without a watch we cannot really tell that the state of the production resource was indeed consistent.